File: LL_palindrome.java
// Time Complexity : O(N)
// Space Complexity : O(h) h is height of recursive stack.
// Did this code successfully run on Leetcode : Yes

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        if(head == null || head.next == null)
            return true;
        ListNode slow = head;
        ListNode fast = head;
        while(fast.next != null && fast.next.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }
        ListNode reverseHead = reverse(slow.next);
        slow.next = null;
        slow = head;
        while(reverseHead != null){
            if(slow.val != reverseHead.val){
                return false;
            }
            slow = slow.next;
            reverseHead = reverseHead.next;
        }
        return true;
    }
    private ListNode reverse(ListNode head){
        if(head == null || head.next == null)
            return head;
        ListNode temp = reverse(head.next);
        head.next.next = head;
        head.next = null;
        return temp;
    }
}

File: llPalindrome.py
#Time Complexity: O(n)
#Space Complexity: O(1)

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        length=0
        slow=head
        fast=head
        while fast!=None and fast.next!=None:
            slow=slow.next
            fast=fast.next.next
        prev=None
        while slow!=None:
            temp=slow.next
            slow.next=prev
            prev=slow
            slow=temp
        while prev:
            if head.val!=prev.val:
                return False
            head=head.next
            prev=prev.next
        return True
        
        
        